### ISSUE TITLE
annotations in JL5AnnotatedElementExt should have setter

### ISSUE THREAD

**Issue by xcv58** [polyglot-compiler/polyglot#3]
In `JL5NodeFactory_c`, it directly set `ext.annotations`
```java
        JL5LocalDeclExt ext = (JL5LocalDeclExt) JL5Ext.ext(n);
        ext.annotations = CollectionUtil.nonNullList(annotations);
```
https://github.com/polyglot-compiler/polyglot/blob/master/src/polyglot/ext/jl5/ast/JL5NodeFactory_c.java#L468

Because the `List<AnnotationElem> annotations` is `protected` in `JL5AnnotatedElementExt`
https://github.com/polyglot-compiler/polyglot/blob/master/src/polyglot/ext/jl5/ast/JL5AnnotatedElementExt.java#L50

But if we extends language based on JL5 extension. It's not possible to set `annotations` without change original jl5 package.

So it's better if we have a public setter function.

**Comment by chinawat** [comment]
Hi Yihong,

There is a workaround in the node factory of an extension.  Since the extension does not have direct access to annotations field, any annotations to be stored must be passed to JL5NodeFactory, as done similarly in that method:

```java
LocalDecl n = super.LocalDecl(pos, flags, annotations, type, name, init);
YourLocalDeclExt ext = ...;
... // set your additional fields here
return n;
```

In general, a setter would not be appropriate for nodes or extensions, because the AST is supposed to be immutable.  The issue you are raising points out a problem regarding the fundamental design of Polyglot: The base language is responsible for creating node and extension objects, but the parent language does not know anything about states in the child language.  As a result, states are left uninitialized and require a backpatch as above.  A better approach would be to construct extension objects from the deepest extension upward, initializing states as we go along, so that we have a completely initialized object by the time we have every node and extension object created.

**Comment by xcv58** [comment]
Hi Chin,

I'm totally agree with you. But the thing is that if we create language based on JL5. Particularly, if we extends `LocalDecl.java`, it requires change `LocalDecl` function in `NodeFactory_c`. Is this ordinary way to use polyglot? If no, please do tell me how to do that correctly.

Otherwise, we have to set `annotations` properly like:
```java
    @Override
    public LocalDecl LocalDecl(Position pos, Flags flags,
            List<AnnotationElem> annotations, TypeNode type, Id name, Expr init) {
        LocalDecl n = super.LocalDecl(pos, flags, type, name, init);
        JL5LocalDeclExt ext = (JL5LocalDeclExt) JL5Ext.ext(n);
        ext.annotations = CollectionUtil.nonNullList(annotations);
        return n;
    }
```

If we don't set `annotations`, it raise `NullPointerException` because `annotations` is `null`.

But out of the jl5 package, we're unable to change `ext.annotations`. So that's why I suggest to add setter.

**Comment by chinawat** [comment]
Hi Yihong,

I will assume that in your extension, you have a new state (field) for a LocalDecl.  If not, the node factory of your extension can simply extend JL5NodeFactory without overriding LocalDecl() factory method, and the problem would be solved.

The idea is that you store everything in the extension object:
```java
class YourLocalDeclExt extends YourExt /* could be something else, depending on your extension hierarchy */ {
    Object yourNewState;
    ...
}
```

Then, you have to define LocalDecl() factory method to set your state, and call the LocalDecl() factory method in the superclass, which sets the annotations:
```java
class YourNodeFactory_c extends JL5NodeFactory_c {

    public LocalDecl LocalDecl(Position pos, Flags flags,
            List<AnnotationElem> annotations, TypeNode type, Id name, Expr init, Object yourNewState) {
        LocalDecl n = super.LocalDecl(pos, flags, annotations, type, name, init);
        YourLocalDeclExt ext = ...;
        ext.yourNewState = ...;
        return n;
    }
}
```
It is also a good idea to override LocalDecl() factory method from JL5 to call your newly declared method with the default value for your new state, and declare that overriding method final.

To recap, we simply follow the same pattern used in JL5NodeFactory for your extension.  Note that we did not extend LocalDecl_c from the base language.

**Comment by xcv58** [comment]
We're almost doing the same thing above.

Here is our implementation: https://github.com/blue-systems-group/project.maybe.polyglot/blob/maybe/maybe/compiler/src/edu/buffalo/cse/blue/maybe/ast/MaybeNodeFactory_c.java#L103

The thing is we wanna our `MaybeLocalDecl`, which extends from `LocalDecl`, can take advantage from `JL5LocalDeclExt`.

The problem is `JL5LocalDeclExt` extends `JL5AnnotatedElementExt`, the `JL5AnnotatedElementExt` has its own state: `annotations` which need to be not `null`. And we cannot simply set it to an empty list.

**Comment by xcv58** [comment]
We'll try to store `states` to our ext. So does that mean we need add our own "MaybeLocalDeclToJL_c" and "MaybeLocalDeclToExt_c"?

**Comment by chinawat** [comment]
It looks like your `MaybeLocalDecl` is an extended version of a local declaration.  I am not sure why you did not extend `LocalDecl_c` to begin with.

At any rate, from what I understand so far, I think you can convert MaybeLocalDecl_c to MaybeLocalDeclExt as above.  It does not hurt to add the two translate classes that you mentioned.  (I am not entirely familiar with the purposes of these classes, though.)

Looking more closely at our code, I see that JL5LocalDeclExt does provide a setter method for annotations, but it is inherited from JL5AnnotatedElementExt.  So, you could retain existing classes if you'd like, but I prefer the pattern I described above better as long as you don't have an AST node that is entirely new to the language (e.g., the extended for (T e : c) in Java 5).  Ideally, we should adhere to just one pattern that avoids setter methods, but the existing design of Polyglot poses issues as discussed.

**Comment by xcv58** [comment]
OK, I think I got it.

We need to refactor our codes to put all logic in our own `ext` class. And override the `prettyPrint` to print ordinary java codes.

I already used the `annotationElems` to avoid change `annotations` to `public`.

I knew there's `public void setAnnotations(Annotations annotations)`. But its parameter is `Annotations` not `List<AnnotationElem> annotations`. So we have to create an Annotations object.

Never mind about this issue. We made mistake when implement our minor syntax change.

Thank you very much! I'm learning a lot about polyglot.

**Comment by chinawat** [comment]
Both ```annotationElems``` methods in JL5AnnonatedElementExt are indeed getter and setter (more correctly, mutator) for ```annotations``` state.  Why they have different names, that I don't know.

Keep in mind that almost everything you can do in base AST classes can be done in extension classes, so you will find that you rarely need to extend any base AST class.

**Comment by xcv58** [comment]
Yeah, I regret that I haven't realize that before.
I'm really appreciated your help!
